add auto-setup blogpost (#804)
temporalio/documentation: blog/2021-12-29-auto-setup.md

---
tags:
  - Temporal
  - Server

> posted_on_: 2021-12-29T00:00:09Z

Author: ```suggestion
posted_on_: 2022-01-05T00:00:09Z
```

slug: auto-setup
title: How Temporal Auto-Setup Works
author: swyx
author_title: Head of Developer Experience
author_image_url: https://avatars.githubusercontent.com/u/6764957?v=4
release_version: V1.14.0
---

> We often say that “Temporal Server is a simple Go binary”, and yet most users are insulated from using this binary directly. There are several layers of abstraction to get through (docker-compose or Helm Chart, Dockerfile, and the auto-setup.sh script) before you reach this baseline understanding of Temporal. Together they comprise the environment requirements of Temporal:

Reviewer: But is this really even true. Things like Temporalite prove you absolutely do not need Docker, Helm or others to have a "baseline understanding" of Temporal.

Author: i think if you read it again i'm saying the exact same thing - the baseline is what's underneath the docker, helm, other stuff

> We often say that “Temporal Server is a simple Go binary”, and yet most users are insulated from using this binary directly. There are several layers of abstraction to get through (docker-compose or Helm Chart, Dockerfile, and the auto-setup.sh script) before you reach this baseline understanding of Temporal. Together they comprise the environment requirements of Temporal:

Reviewer 2: ```suggestion
We often say that “Temporal Server is a simple Go binary”, and yet most users are insulated from using this binary directly. There are several layers of abstraction to get through (docker-compose or Helm Chart, Dockerfile, and the `auto-setup.sh` script) before you reach this baseline understanding of Temporal. Together they comprise the environment requirements of Temporal:
```

<!--truncate-->

![image](https://user-images.githubusercontent.com/6764957/147678999-883be1b4-4d32-4c89-84a4-00e8b701cdef.png)

> While the other pieces are industry standard formats with explanations available elsewhere, the Auto-Setup is the most proprietary of them, and so essential that we often recommend users just use the [temporalio/auto-setup](https://hub.docker.com/r/temporalio/auto-setup) Docker image, over our Docker image with Temporal Server itself (most beginners don’t even know this because we abstract it with docker-compose). 

Reviewer: Why is Auto Setup caps?

> While the other pieces are industry standard formats with explanations available elsewhere, the Auto-Setup is the most proprietary of them, and so essential that we often recommend users just use the [temporalio/auto-setup](https://hub.docker.com/r/temporalio/auto-setup) Docker image, over our Docker image with Temporal Server itself (most beginners don’t even know this because we abstract it with docker-compose). 

Reviewer 2: ```suggestion
While the other pieces are industry standard formats with explanations available elsewhere, the auto-setup is the most proprietary of them, and so essential that we often recommend users just use the [temporalio/auto-setup](https://hub.docker.com/r/temporalio/auto-setup) Docker image, over our Docker image with Temporal Server itself (most beginners don’t even know this because we abstract it with docker-compose). 
```

It’s worth: 

> - knowing that the `auto-setup.sh` script exists and is an important part of the Temporal deployment model,

Reviewer: weird trailing comma

Author: ```suggestion
- knowing that the `auto-setup.sh` script exists and is an important part of the Temporal deployment model
```

> - understanding what it does to fill in the blanks between “it’s just a binary, any decent dev can figure it out from here right?” to production ready cluster, and

Reviewer: trailing and

Author: ```suggestion
- understanding what it does to fill in the blanks between “it’s just a binary, any decent dev can figure it out from here right?” to production ready cluster
```

- understanding what is optional so you can modify it when it is getting in your way, or have confidence throwing it out and writing your own.

> The goal of this post is ***not*** to explain every little detail (you can [read the code](https://github.com/temporalio/temporal/blob/master/docker/auto-setup.sh) yourself for that), but to highlight important pieces every self-hosted Temporal user should know as essential system knowledge for deploying Temporal Server.

Reviewer: the part after the comma is a mouthful to read.

Author: ```suggestion
The goal of this post is ***not*** to explain every little detail (you can [read the code](https://github.com/temporalio/temporal/blob/master/docker/auto-setup.sh) yourself for that), but to highlight important pieces every self-hosted Temporal user should know.
```

## Auto-Setup at a High Level

Our [basic Docker entrypoint](https://github.com/temporalio/temporal/blob/master/docker/entrypoint.sh) does some bindings, but most importantly, it runs `[auto-setup.sh](https://github.com/temporalio/temporal/blob/master/docker/auto-setup.sh)` then `[start-temporal.sh](https://github.com/temporalio/temporal/blob/master/docker/start-temporal.sh)`.  

`auto-setup.sh` does a few things (we will go deeper on the **bolded** parts):

1. Temporal Schema Setup (if you don’t opt out):

>     1. Validates seed environment variables (trivial step)

Reviewer: Is this really worth putting if its trivial

Author: i like giving indication of relative size of steps from the 50k foot view

    2. Waits for the database to start (using `nc` for SQL databases, or [`temporal-cassandra-tool`](https://github.com/temporalio/temporal/blob/master/tools/cassandra/README.md) for Cassandra)
    3. **Setup Temporal Schema on that database (more on this below)**
2. Elasticsearch Schema Setup (if you opt in):
    1. Validates seed environment variables (trivial step)
    2. Waits for Elasticsearch to start (optionally, timed out with a `ES_SCHEMA_SETUP_TIMEOUT_IN_SECONDS` env var)
    3. Setup Temporal’s Enhanced Visibility Schema on that Elasticsearch instance
3. Setup Temporal Server:
    1. Wait for Temporal Server to start (from running `[start-temporal.sh](http://start-temporal.sh)` concurrently, and checked using `tctl cluster health | grep SERVING`
    2. **Register the `default` namespace** (unless opted out with `SKIP_DEFAULT_NAMESPACE_CREATION`)

>     3. Add basic custom search attributes (unless opted out with `SKIP_ADD_CUSTOM_SEARCH_ATTRIBUTES`)

Reviewer: Are these the "semi-custom" attributes? If they aren't, why is it required step (there is no caveat about users needing to specify their own custom attributes.

Author: i dont know either, this was a surprise to find when i went thru the code. @alexshtin can perhaps offer more info here. i suspect it is just historical artifact

> - Note that this script only handles setup - it doesn’t actually run the `temporal-server` binary itself. That job is given to `start-temporal.sh`, whose main job is to run `temporal-server` with flags for given service roles (mainly specified by the `SERVICES` env var).

Reviewer: Normally you would handle setting up schemas differently too before this step (as far as I know)

Author: so far i havent seen any guidance on how to do this for prod...  frankly am impressed that people figure this stuff out on their own

- This is the namespace that most beginners are used to seeing when they start up Temporal inside docker-compose - because the `auto-setup.sh` script does it for them! Without this script running, Temporal Server doesn’t have any namespaces registered. This can come as a surprise to some users who assume the default namespace is a necessary part of Temporal.
- All SDKs Workers and Clients, as well as Temporal Web, connect to the `default` namespace if not specified - most production deployments of Temporal will want to connect to specific namespaces. (try running [temporalite](https://github.com/DataDog/temporalite) without any namespaces to see how SDK code without namespaces specified stops working!)
- The script allows you to customize the name and retention period with the `DEFAULT_NAMESPACE` and `DEFAULT_NAMESPACE_RETENTION` env vars if you wish, but a production setup may want to register a number of namespaces differently.

> The final part of the `auto-setup.sh` script also does registers some basic custom search attributes:

Reviewer: ```suggestion
The final part of the `auto-setup.sh` script also does registers some pre-defined custom search attributes:
```

(only when running with Elasticsearch I believe.

```tsx
tctl --auto_confirm admin cluster add-search-attributes \
          --name CustomKeywordField --type Keyword \
          --name CustomTextField --type Text \
          --name CustomIntField --type Int \
          --name CustomDatetimeField --type Datetime \
          --name CustomDoubleField --type Double \
          --name CustomBoolField --type Bool
```

> But these are for demo purposes (eg [for code samples](https://github.com/temporalio/samples-go/blob/77728cf7c38570898b2c90bf6eb0720c7f5fb30d/searchattributes/searchattributes_workflow.go#L56-L63)) rather than illustrative of real usage - hence the extremely generic naming of these attributes. In practice, you are likely to drop this in favor of manual or custom setup of search attributes.

Reviewer: Don't actually think thats true. These are nice because there is a limit on search attributes and its a pain to add new ones for some people.

Author: ```suggestion
These are intentionally generically named, for demo purposes (eg [for code samples](https://github.com/temporalio/samples-go/blob/77728cf7c38570898b2c90bf6eb0720c7f5fb30d/searchattributes/searchattributes_workflow.go#L56-L63)) and for ease of use (since they are already set up, you don't have to add them when you find you need them later). As you advance in your usage, you may wish to drop this step in favor of better named and specified attributes - just keep in mind that you are [limited to a maximum of 100](https://docs.temporal.io/docs/server/production-deployment/#server-limits).
```
